kernel/src/hil/kv_store.rs

```diff
@@ -0,0 +1,84 @@
+//! Interface for KV Stores
+
+use crate::returncode::ReturnCode;
+
+#[derive(Clone, Copy, PartialEq)]
+pub enum Operation {
+    None,
+    Init,
+    GetKey,
+    AppendKey,
+    InvalidateKey,
+    GarbageCollect,
+}
+
+/// The type of keys, this should define the output size of the digest
+/// operations.
+pub trait KeyType: Eq + Copy + Clone + Sized + AsRef<[u8]> + AsMut<[u8]> {}
+
+impl KeyType for [u8; 64] {}
+
+/// Implement this trait and use `set_client()` in order to receive callbacks.
+pub trait Client<'a, K: KeyType> {
+    /// This callback is called when any operation is complete.
+    ///
+    /// 'operation': The operation that completed
+    /// 'result': Nothing on success, 'ReturnCode' on error
+    /// 'key': A key buffer is one exists
+    /// 'value': A value buffer is one exists
+    /// 'buf': A buf buffer is one exists
+    fn operation_complete(
+        &'a self,
+        operation: Operation,
+        result: Result<(), ReturnCode>,
+        key: Option<&'static mut K>,
+        value: Option<&'static mut [u8]>,
+        buf: Option<&'static mut [u8]>,
+    );
+}
+
+pub trait KVStore<K: KeyType> {
```

Comment:
Using generics is probably fine, but I think that it's more idiomatic in the Rust universe to use an associated type in such a case. It seems close to the examples from the [_Rust by example_ section](https://doc.rust-lang.org/rust-by-example/generics/assoc_items/types.html).

As per this [StackOverflow answer](https://stackoverflow.com/a/32065644) it also makes sense to ensure that there can be only one implementation of `KVStore` on a given type, not multiple (which an associated type would assure). I quite like that you don't have to write out the full type specification every time (as it is unambiguous then).

Reply:
Changed.

Reply:
Now this seems to use both generics and an associated type -- I think you want to have this be just 
```rust
pub trait KVStore {
    type K: KeyType;
    //...
}
```

Reply:
Yes, it shouldn't be mixed in one trait. Although I think it's good that the client still uses generics. A consumer implementing it might want to access multiple key-value stores with different `KeyType`s, which will be possible with generics.

Reply:
Whoops, fixed.